packages/techdocs-cli/src/lib/run.ts

```diff
@@ -0,0 +1,106 @@
+/*
+ * Copyright 2020 The Backstage Authors
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+import { spawn, SpawnOptions, ChildProcess } from 'child_process';
+
+export type LogFunc = (data: Buffer | string) => void;
+type SpawnOptionsPartialEnv = Omit<SpawnOptions, 'env'> & {
+  env?: Partial<NodeJS.ProcessEnv>;
+  // Pipe stdout to this log function
+  stdoutLogFunc?: LogFunc;
+  // Pipe stderr to this log function
+  stderrLogFunc?: LogFunc;
+};
+
+// TODO: Accept log functions to pipe logs with.
+// Runs a child command, returning the child process instance.
+// Use it along with waitForSignal to run a long running process e.g. mkdocs serve
+export const run = async (
+  name: string,
+  args: string[] = [],
+  options: SpawnOptionsPartialEnv = {},
+): Promise<ChildProcess> => {
+  const { stdoutLogFunc, stderrLogFunc } = options;
+
+  const env: NodeJS.ProcessEnv = {
+    ...process.env,
+    FORCE_COLOR: 'true',
+    ...(options.env ?? {}),
+  };
+
+  // Refer: https://nodejs.org/api/child_process.html#child_process_subprocess_stdio
+  const stdio = [
+    'inherit',
+    stdoutLogFunc ? 'pipe' : 'inherit',
+    stderrLogFunc ? 'pipe' : 'inherit',
+  ] as ('inherit' | 'pipe')[];
+
+  const childProcess = spawn(name, args, {
+    stdio: stdio,
+    shell: true,
+    ...options,
+    env,
+  });
+
+  if (stdoutLogFunc && childProcess.stdout) {
+    childProcess.stdout.on('data', stdoutLogFunc);
+  }
+  if (stderrLogFunc && childProcess.stderr) {
+    childProcess.stderr.on('data', stderrLogFunc);
+  }
+
+  return childProcess;
+};
+
+// Block indefinitely and wait for a signal to kill the child process(es)
+// Throw error if any child process errors
+// Resolves only when all processes exit with status code 0
+export async function waitForSignal(
+  childProcesses: Array<ChildProcess>,
+): Promise<void> {
+  const promises: Array<Promise<void>> = [];
+
+  childProcesses.forEach(childProcess => {
+    if (typeof childProcess.exitCode === 'number') {
+      if (childProcess.exitCode) {
+        throw new Error(`Non zero exit code from child process`);
+      }
+      return;
+    }
+
+    for (const signal of ['SIGINT', 'SIGTERM'] as const) {
```

Comment:
@andrewthauer can you check this PR https://github.com/backstage/techdocs-cli/pull/186 that @camilaibs contributed to fix the processes not being killed and exit the parent process only when all child promises are resolved. Specific lines of code https://github.com/backstage/techdocs-cli/pull/186/files#diff-6081b034129b693eea1126455b3b9fc81b19b3f2436f824542b0ca91a61fde3dR74-R81. I think that can help us get the e2e test to pass which is the last thing before getting this 🚢

Reply:
Sorry, I've been really busy with no time to fiddle with trying to get the e2e tests working. Looks like @camilaibs is experimenting a bit? I'm happy to hand this off if someone has a bit more time to dedicate to getting this to the finish line.

Reply:
Hey @emmaindal and @andrewthauer, I've tried [here](https://github.com/backstage/backstage/pull/8007) on another branch built on this and it looks like all the checks, including end-to-end testing, are passing . If you guys have time, it would be great to test everything again before merging, just to make sure nothing broke after my changes.

Reply:
Awesome! Thanks both of you, I'll test it and review it today!